src/Middleware/Router.php

```diff
@@ -58,4 +58,49 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
             ->withDispatcher($this->dispatcher)
             ->process($request, $handler);
     }
+
+    public function ignoreMethodFailureHandle(): self
+    {
+        $new = clone $this;
+        $new->ignoreMethodFailureHandle = true;
+        return $new;
+    }
+
+    public function withOptionsResponseFactory(MethodsResponseFactoryInterface $optionsResponseFactory): self
+    {
+        $new = clone $this;
+        $new->optionsResponseFactory = $optionsResponseFactory;
+        return $new;
+    }
+
+    public function withNotAllowedResponseFactory(MethodsResponseFactoryInterface $notAllowedResponseFactory): self
+    {
+        $new = clone $this;
+        $new->notAllowedResponseFactory = $notAllowedResponseFactory;
+        return $new;
+    }
+
+    /**
+     * @param string[] $methods
+     */
+    private function getOptionsResponse(ServerRequestInterface $request, array $methods): ResponseInterface
+    {
+        return $this->optionsResponseFactory !== null
+                ? $this->optionsResponseFactory->create($methods, $request)
+                : $this->responseFactory
+                    ->createResponse(Status::NO_CONTENT)
+                    ->withHeader(Header::ALLOW, implode(', ', $methods));
+    }
+
+    /**
+     * @param string[] $methods
+     */
+    private function getMethodNotAllowedResponse(ServerRequestInterface $request, array $methods): ResponseInterface
+    {
+        return $this->notAllowedResponseFactory !== null
+                ? $this->notAllowedResponseFactory->create($methods, $request)
+                : $this->responseFactory
+                    ->createResponse(Status::METHOD_NOT_ALLOWED)
+                    ->withHeader(Header::ALLOW, implode(', ', $methods));
```

Comment:
Extract it to new response factory class and use it by default for `notAllowedResponseFactory`.